### PR TITLE
Game name autocompletion added to search bar

### LINKED ISSUE #95: Create a functional search bar with autocompletion for game names
Add a search bar that will autocomplete a user's query to search for games registered in ChiGame. We will need to wait on the user stories for searching to be completed in issue #90. The research for how to create a functional search bar was done in issue #91. The research for how to implement autocomplete was done in issue #94. This issue involves integrating that knowledge into one search bar with autocomplete. This search bar may share functionality with the users team. (Edited by Patrick)

### PR DESCRIPTION
**Pull request by ptlin25**
The search bar now autocompletes game names when the Game option is selected in the dropdown menu. The search bar finds the first 10 autocompletion results and displays the tags underneath the search bar. Clicking on the tag will redirect the user to the appropriate page. If a different search option is selected, the autocompletion is disabled for now. This can be changed in the future to autocomplete the relevant tags.

### CODE REVIEW

File: src/templates/base.html [hunk 1]
```diff
@@ -136,23 +142,91 @@
...
+                request.term is the string entered into the search bar
+                "i" means that the matching is case insensitive
+                */
+                var matcher = new RegExp($.ui.autocomplete.escapeRegex(request.term), "i");
+                // results is all of the game names that match the term in the search bar
+                var results = $.grep(game_names, function(tag){
+                  return matcher.test(tag.label)
+                })
+                // only display the first 10 results
+                response(results.slice(0, 10));
```
Comment (rakrusch) [inline src/templates/base.html]: Minor unnecessary stylistic idea, but we could consider creating a constant that we set equal to 10 and have that constant here to improve readability and make it easier if we ever want to change the number of results we are displaying.

File: src/templates/base.html [hunk 2]
```diff
@@ -136,23 +142,91 @@
...
-      <script>
+      {# disable djlint for this section because the linter breaks the template tags #}
+      <!-- djlint:off --><script>   // "djlint:off" turns the linter off for this section of code
         document.addEventListener('DOMContentLoaded', function() {
           const search_bar = document.getElementById('search-bar');
           const query_type = document.getElementById('query-type');
 
-          query_type.addEventListener("change", function() {
+          // by default, query autocompletes game names
+          {% get_games as games %}
```
Comment (rakrusch) [inline src/templates/base.html]: I really like your use of comments to explain the purpose of code! If you want to more detailed with comments, you could related to important logical steps. For example, here you could add a comment next to 
{% get_games as games %}
 stating that we are:
// Fetching game names for autocomplete

Though, this is not strictly necessary and your use of comments already is really good.

File: src/templates/base.html [hunk 3]
```diff
@@ -136,23 +142,91 @@
...
+                    request.term is the string entered into the search bar
+                    "i" means that the matching is case insensitive
+                    */
+                    var matcher = new RegExp($.ui.autocomplete.escapeRegex(request.term), "i");
+                    // results is all of the game names that match the term in the search bar
+                    var results = $.grep(game_names, function(tag){
+                      return matcher.test(tag.label)
+                    })
+                    // only display the first 10 results
+                    response(results.slice(0, 10));
```
Comment (rakrusch) [inline src/templates/base.html]: Same idea here with using a constant

**Review by rakrusch — APPROVED** [review]
This looks really good Patrick! I really like your use of inline comments to make code much more readable. I made some small comments but will approve this PR anyway, as the changes I suggest are minor and not strictly necessary.

File: src/templates/base.html (hunk 1 above)
Comment (majorsylvie) [reply src/templates/base.html]: yes I agree 100%, avoiding magic numbers and instead having this be stored as some `AUTOCOMPLETION_RESULTS_LIMIT` constant.

File: src/chigame/games/templatetags/tags.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from django import template
+
+from chigame.games.models import Game
+
+register = template.Library()
```
Comment (majorsylvie) [inline src/chigame/games/templatetags/tags.py:8]: Please add comments explaining what is happening in this code here. 
As template tags are something not covered in the course IIRC, thus explaining more of this code would be good for the rest of the chigame devs.

File: src/chigame/games/templatetags/tags.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+
+@register.simple_tag
+def get_games():
```
Comment (majorsylvie) [inline src/chigame/games/templatetags/tags.py:13]: Similar to above, add a comment explaining why this code is here and what purpose it serves.

File: src/templates/base.html (hunk 2 above)
Comment (majorsylvie) [reply src/templates/base.html]: I agree, adding an explicit comment saying line 182 is a query would be quite nice.

File: src/templates/base.html [hunk 6]
```diff
@@ -171,17 +178,83 @@
...
+                label: "{{game.name}}"
+              },
+            {% endfor %}
+          ];
+          $("#query").autocomplete({
+              source: function(request, response) {
+                /*
+                new regular expression object (RegExp) to search for pattern
+                request.term is the string entered into the search bar
+                "i" means that the matching is case insensitive
```
Comment (majorsylvie) [inline src/templates/base.html:192]: great inclusion with this `'i'` comment

File: src/templates/base.html [hunk 7]
```diff
@@ -171,17 +178,83 @@
...
+                disabled: true  // turn off autocomplete for users option
+              });
             }
           });
         });
-      </script>
-    {% endblock inline_javascript %}
-  </body>
+</script><!-- djlint:on -->
+{# turn djlint back on#}
```
Comment (majorsylvie) [inline src/templates/base.html]: I would add a comment about why this is necessary!

**Review by majorsylvie — CHANGES_REQUESTED** [review]
The functionality of no-autocomplete while selecting the `Users` from dropdown does not appear to work:

https://github.com/uchicago-cs/chigame/assets/86740254/16a285c0-3337-4d6c-8f69-39ec333c360d


Besides that, the actual autocompletion is wonderful and look genuinely good. I appreciate that you are directly linked to the page and the choice to limit to the top 10 results. 

Please add the code comments and constant, then actually make sure that you're dropdown-switching works

**Comment by ptlin25** [comment]
Thanks for catching the issue with the autocompletion of game names even when the user option is selected. I have fixed that issue and also added more comments to explain the code.

File: src/chigame/games/templatetags/tags.py (hunk 4 above)
Comment (majorsylvie) [reply src/chigame/games/templatetags/tags.py:8]: great comment added!

**Review by majorsylvie — APPROVED** [review]
Now with the forum dropdown merged in #309, this code still works and only autocompletes for the games search!

Great work!

One thing to note, you may want to consider some kind of search ranking that prioritizes the beginnings of words / titles.
Since here searching with `"chi"` returns `Luftschiff` before `Chicago`, which seems odd.

Doesn't have to be a change made in this PR, **please add an issue to the backlog to do this if you feel it's possible!**
image

**Comment by ptlin25** [comment]
When the search bar in the navigation bar has the "Games" option in the dropdown menu selected, the search bar will autocomplete game names in the database. This is done through the jQuery UI Autocomplete widget. A custom template tag was added to get a list of all of the games in the chigame database. This list is used to create the autocomplete tags. After a user enters a phrase into the search bar, the first 10 autocomplete results are shown. If the user clicks on the autocomplete options, they will be redirected to the corresponding game's detail page. Currently, there is no autocomplete for other search bar modes. More info can be found on the [wiki page](https://github.com/uchicago-cs/chigame/wiki/Games-~-Game-Name-Autocompletion).

**Comment by majorsylvie** [comment]
Issue Score: Excellent

Comments:
Phenomenal work! This is an awesome PR that was very thoroughly done. 
I appreciate the great closing comments, good reviews (thanks @rakrusch)

Be very proud of this great work! We're almost done with the quarter!

**Pull request merged**